src/openfermion/measurements/_probability_distributions.py

```diff
@@ -0,0 +1,378 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Classes to store probability distributions for estimation
+of random variables.
+"""
+import numpy
+from scipy import sparse
+from numpy import cos, sin, pi
+from openfermion.config import EQ_TOLERANCE
+
+
+class FourierProbabilityDist(object):
+    """
+    Stores an approximate probability distribution of our knowledge of
+    a quantum state in the eigenbasis of a unitary operator.
+
+    A generic quantum state may be written in the eigenbasis of a unitary
+    operator U,
+
+    $|\psi\rangle = sum_j a_j |phi_j\rangle$,
+
+    where psi is the quantum state, |phi_j\rangle are the eigenstates of U,
+    labeled by their eigenvalues ($U|phi_j\rangle = phi_j|phi_j\rangle$),
+    and a_j are the amplitudes.
+
+    This class stores a distribution P(phi_0, phi_1,... , A_0, A_1,...)
+    (where A_j = |a_j|^2) of our knowledge of these variables which may
+    then be updated by the results of experiments (in particular here,
+    QPE experiments).
+
+    This class approximates this distribution as:
+    P(phi_0, phi_1,... , A_0, A_1,...)=P(A_0, A_1,...)*\prod_jP(\phi_j).
+
+    It then represents each phase distribution:
+    P(\phi_j) = sum_n (c_{j,2n} cos(n*phi_j) + c_{j,2n-1} sin(n*phi_j))
+
+    Separately, it stores an estimate of the mean amplitude \bar{A_j},
+    and data that allows for this to be updated (rather than a direct
+    representation of the function).
+
+    More details can be found in arXiv:1809.09697, Appendix C.
+    """
+
+    def __init__(self,
+                 amplitude_mean=None,
+                 amplitude_var=None,
+                 num_vectors=1,
+                 num_freqs=1000,
+                 fourier_vector=None):
+        """
+        Args:
+            amplitude_mean (numpy array or list or None):
+                We take an normal distribution for the amplitudes
+                which requires for each a mean and a variance.
+                We only allow None for a single vector (in which case
+                the amplitude is fixed to 1), as we need this
+                to break the symmetry of the problem.
+            amplitude_var (numpy array or list or None): variance of the
+                estimated distribution in amplitude_mean.
+                We only allow None for a single vector (in which case
+                this variable is unused).
+            num_vectors (int): number of phases phi_j to estimate
+            num_freqs (int): number of wave components cos(n*phi_j)
+                to store coefficients for (this provides a cut-off to
+                the sensitivity of the function).
+            fourier_vector (numpy array or None): the function
+                given in terms of the Fourier components c_{j,n}.
+                If None, prior taken as a flat function over [-pi,pi].
```

Comment:
Please document the expected shapes of all the numpy array arguments.

Reply:
Done